packages/flutter_tools/lib/src/base/process_manager.dart

```diff
@@ -479,3 +491,294 @@ class _RecordingProcess implements Process {
+
+  @override
+  final dynamic stderr;
+
+  _ReplayProcessResult._({this.pid, this.exitCode, this.stdout, this.stderr});
+
+  static Future<_ReplayProcessResult> create(
+    Directory dir,
+    Map<String, dynamic> entry,
+  ) async {
+    String basePath = path.join(dir.path, entry['basename']);
+    return new _ReplayProcessResult._(
+      pid: entry['pid'],
+      exitCode: entry['exitCode'],
+      stdout: await _getData('$basePath.stdout', entry['stdoutEncoding']),
+      stderr: await _getData('$basePath.stderr', entry['stderrEncoding']),
+    );
+  }
+
+  static Future<dynamic> _getData(String path, String encoding) async {
+    File file = new File(path);
+    assert(await file.exists());
+    return encoding == null
+        ? await file.readAsBytes()
+        : await file.readAsString(encoding: _getEncodingByName(encoding));
+  }
+
+  static _ReplayProcessResult createSync(
+    Directory dir,
+    Map<String, dynamic> entry,
+  ) {
+    String basePath = path.join(dir.path, entry['basename']);
+    return new _ReplayProcessResult._(
+      pid: entry['pid'],
+      exitCode: entry['exitCode'],
+      stdout: (_getDataSync('$basePath.stdout', entry['stdoutEncoding'])),
+      stderr: (_getDataSync('$basePath.stderr', entry['stderrEncoding'])),
+    );
+  }
+
+  static dynamic _getDataSync(String path, String encoding) {
+    File file = new File(path);
+    assert(file.existsSync());
+    return encoding == null
+        ? file.readAsBytesSync()
+        : file.readAsStringSync(encoding: _getEncodingByName(encoding));
+  }
+
+  static Encoding _getEncodingByName(String encoding) {
+    if (encoding == 'system')
+      return const SystemEncoding();
+    else if (encoding != null)
+      return Encoding.getByName(encoding);
+    return null;
+  }
+
+  Process asProcess(bool daemon) {
+    assert(stdout is List<int>);
+    assert(stderr is List<int>);
+    return new _ReplayProcess(this, daemon);
+  }
+}
+
+/// A [Process] implementation derives its data from a recording fragment.
+class _ReplayProcess implements Process {
+  @override
+  final int pid;
+
+  final List<int> _stdout;
+  final List<int> _stderr;
+  final StreamController<List<int>> _stdoutController;
+  final StreamController<List<int>> _stderrController;
+  final int _exitCode;
+  final Completer<int> _exitCodeCompleter;
+
+  _ReplayProcess(_ReplayProcessResult result, bool daemon)
+      : pid = result.pid,
+        _stdout = result.stdout,
+        _stderr = result.stderr,
+        _stdoutController = new StreamController<List<int>>.broadcast(),
+        _stderrController = new StreamController<List<int>>.broadcast(),
+        _exitCode = result.exitCode,
+        _exitCodeCompleter = new Completer<int>() {
+    // Produce stream events after a delay so that all interested parties have
+    // a chance to add themselves as listeners before we fire our stream events
+    // and terminate the process.
+    new Timer(const Duration(milliseconds: 50), () {
```

Comment:
Ah... maybe this delay before adding data to stdout/stderr alleviates the issue I raised above. I prefer not to have arbitrary delays in the code, but this should work.